woocommerce/woocommerce-gutenberg-products-block
# Add cart/extensions endpoint to allow third party extensions to process server-side and get a new cart

This PR is based on `feature/coupon-area-slot-fill` as it forms part of the work I'm doing for WooCommerce Points and Rewards.

--- 

This PR will add the ability for extensions to register functions/callbacks to be executed by Blocks when the `cart/extensions` endpoint is hit.

This PR is necessary to allow third-party extensions to tell Blocks they need to do some server-side processing which will update the cart.

The way it should work is an extension will register its callbacks with Blocks, via the `register_update_callback` method. When the `wc/store/cart/extensions` endpoint is requested, there should be at the minimum, a `namespace` and `endpoint` parameter passed, as well as any additional data. If any callbacks are registered, matching the `namespace` and `endpoint` values, they are executed, and then the cart is returned by the API.

 ---

The changes made in this PR are:
- Delete `updateCartFromAPI` JS function and replace it with `extensionCartUpdate` - this is a JS function intended for use by extensions to signal to Blocks that it should POST to the `/cart/extensions` endpoint, then pass the result to `receiveCart`. This function returns a promise so extensions can know when the request is finished processing.
- The args passed to this function should be: 
```typescript
{
  namepsace: string,
  data: object,
}
```
- Add the `cart/extensions` route. Extensions may signal to blocks that they need something to be done server-side which will result in a modified cart, and Blocks will make the request on their behalf.
- Add two methods to `src/Domain/Services/ExtendRestApi.php`: `register_update_callback` and `get_update_callback`. Extensions will use `register_update_callback` to tell Blocks which `callable`s to execute when the `cart/extensions` endpoint is hit. The `get_update_callback` will be used internally only, to determine which callbacks ought to be executed, based on the passed namespace and endpoint.
- Add `CartExtensionsSchema` - This is added to allow us to control the flow of the programme and to execute the callbacks before returning the cart.
- Add PHPUnit tests for these new endpoints and test their error handling.

Fixes #4254




### How to test the changes in this Pull Request:

1. In an extension that extends the REST API, you could add
```php
self::$extend->register_update_callback(
  array(
    'namespace'         => self::IDENTIFIER,
    'callback_function' => array( 'WC_Points_Rewards_Cart_Checkout', 'rest_apply_discount' ),
  )
);
```
Changing the `callback_function` to a callable of your choice. Subscriptions might be a good one to test with.
2. On the front-end use the `extensionCartUpdate` function to call the endpoint, and ensure the callable you registered with `register_update_callback` is executed.



### Changelog

> Add the ability for extensions to register callbacks to be executed by Blocks when the cart/extensions endpoint is hit. Extensions can now tell Blocks they need to do some server-side processing which will update the cart.
